### ISSUE TITLE
feat(engine): backend protocol — multi-backend evaluation support

### ISSUE THREAD

**Issue by pistore** [fbk-most/civic-digital-twins#171]
\`Evaluation.evaluate()\` is currently coupled to numpy at three levels:

1. **Callable binding** — \`NumpyBackend.adapt(fn)\` wraps user callables for the numpy executor. A second backend (JAX, PyTorch) would need \`JaxBackend.adapt(fn)\`, but there is no shared \`Backend\` contract that \`evaluate()\` can dispatch through.

2. **Graph execution** — \`executor.State\` and \`executor.evaluate_nodes()\` traverse the computation graph and dispatch each node to numpy operations (\`np.add\`, \`np.where\`, etc.). This is entirely numpy-specific.

3. **Result shaping** — \`Evaluation.evaluate()\` builds substitution dicts, normalises shapes, and marginalises results using \`np.asarray\`, \`np.reshape\`, \`np.average\`, \`np.multiply.outer\`, etc. These calls are inlined in the evaluation loop.

## Current state

\`NumpyBackend\` exists as a concrete class with \`adapt(fn) → Functor\` and is the only backend. \`evaluate()\` accepts a \`backend=NumpyBackend\` parameter but validates only that it is \`NumpyBackend\`; all three coupling points above remain numpy-specific.

## Work needed

Define a \`Backend\` protocol that covers all three coupling points:

\`\`\`python
class Backend(Protocol):
    type Array  # np.ndarray, jax.Array, torch.Tensor, …

    @classmethod
    def adapt(cls, fn: Callable[..., Array]) -> Functor: ...

    @classmethod
    def make_state(cls, subs: dict[Node, Array], functions: dict[str, Functor]) -> Any: ...

    @classmethod
    def evaluate_nodes(cls, state: Any, *nodes: Node) -> None: ...

    # Array primitives used in Evaluation.evaluate()
    @classmethod
    def asarray(cls, x) -> Array: ...
    @classmethod
    def reshape(cls, x: Array, shape: tuple) -> Array: ...
    @classmethod
    def average(cls, x: Array, weights: Array, axis: int) -> Array: ...
    # … further primitives as needed
\`\`\`

- Make \`NumpyBackend\` a conforming implementation of this protocol.
- Replace all inline \`np.*\` calls in \`Evaluation.evaluate()\` with \`backend.*\` calls.
- Validate the abstraction with a second backend or a test-double backend.
- Remove \`LambdaAdapter\` once all call sites have migrated to \`NumpyBackend.adapt()\` (it was deprecated in the commit that introduced \`NumpyBackend\`).

## Sequencing note

This can proceed independently of per-variant function registries (#135). When graph regions (#136, #137) are introduced, per-region function namespaces and backend selection could merge into a single piece of region metadata — eliminating the need to pass \`backend=\` explicitly at each \`evaluate()\` call.

## Acceptance criteria

- [ ] \`Backend\` protocol defined and documented
- [ ] \`NumpyBackend\` is a conforming implementation
- [ ] \`Evaluation.evaluate()\` delegates all array ops through \`backend\`
- [ ] A second backend or test-double validates the abstraction works
- [ ] \`LambdaAdapter\` removed
